### ISSUE TITLE
Benchmark Idea

### ISSUE THREAD

**Issue by mehdi-haji** [mehdi-haji/DocStill#79]
Proposal: We want to Develop a Benchmark for DocStill Project

Problem Statement

As you know, The DocStill project aims to extract structured knowledge from books, enabling AI agents to automatically generate code from this information, such as implementing trading ideas from textbooks. 

As an AI-heavy application, DocStill requires a robust benchmark to evaluate its knowledge extraction capabilities, assess system quality, and track progress, which is currently lacking. 1
Mandate for Benchmark Development
To address this, two main directives are proposed:
Research Existing Benchmarks: Conduct deep research to identify any similar or exact benchmarks for knowledge extraction from books, understanding how other researchers have tackled comparable challenges. The findings should be summarized. 2
Explore Custom Benchmark: Investigate creating a proprietary benchmark, starting with a data structures and algorithms book. The idea is to use an AI tool like Claude Code to generate "ground truth" data, such as a table of contents or query-answer mappings (e.g., "all sorting algorithms" leading to "Merge sort"). This AI-generated data can then be validated by a subject matter expert to bootstrap a benchmark dataset. 3

What you should deliver:

All findings from both research and custom benchmark exploration should be consolidated into a GitHub issue.

Please use the notify skill to give me updates

**Comment by mehdi-haji** [comment]
# DocStill Benchmark Research & Custom Benchmark Design

## Part 1: Research — Existing Benchmarks for Knowledge Extraction from Books

### 1. Academic Benchmarks for Document Understanding & Long-Document QA

| Benchmark | Source | What It Measures | Size | Metrics | DocStill Relevance |
|-----------|--------|-----------------|------|---------|-------------------|
| **QASPER** | Allen AI (ACL 2021) | QA over full NLP papers; questions from title/abstract, answers require full paper | 5,049 Qs over 1,585 papers | F1, accuracy, ROUGE | **High** — technical doc extraction |
| **NarrativeQA** | DeepMind (ACL 2018) | Reading comprehension over full books (~60K tokens) | 46,765 QA pairs, 1,567 docs | BLEU, ROUGE, F1 | Medium — full-book but narrative |
| **QuALITY** | NYU-MLL (NAACL 2022) | MCQ requiring full-document reading (~5K tokens) | 2,500+ questions | Accuracy | Medium — deep reading needed |
| **SCROLLS** | Tau-NLP (EMNLP 2022) | Suite of 7 long-doc tasks (3K-50K tokens): summarization, QA, MCQ | Aggregation of 7 datasets | Task-dependent (ROUGE, F1, acc) | Medium-High |
| **SQuAD 1.1/2.0** | Stanford NLP | Reading comprehension, paragraph-level | 107K QA pairs | EM, F1 | Low-Medium — paragraph only |
| **BookQA** | Amazon Science (EMNLP 2019) | QA over full-length books using NarrativeQA corpus | 3,427 QA pairs, 614 books | Accuracy, BLEU, ROUGE | Medium |

### 2. Document Parsing Benchmarks

| Benchmark | Source | What It Measures | Size | Metrics | DocStill Relevance |
|-----------|--------|-----------------|------|---------|-------------------|
| **OmniDocBench** | CVPR 2025 | PDF parsing: text, tables, formulas, reading order across 9 doc types incl. textbooks | 1,355 pages, 20K+ annotations | Component-specific metrics | **Very High** — Docling evaluated on this |
| **READoc** | ACL 2025 Findings | PDF → Markdown conversion quality with semantic structure | 3,576 real-world docs | Segmentation-based scoring | **Very High** — exactly DocStill's parsing output |
| **DocLayNet** | IBM Research (KDD 2022) | Layout analysis across diverse domains, 11 element types | 80,863 pages, 6 categories | mAP | **Very High** — Docling's training data |
| **PubTabNet** | IBM (ECCV 2020) | Table structure recognition → HTML | 568K table images | TEDS (Tree-Edit-Distance Similarity) | **High** — tables critical in tech books |
| **PubLayNet** | IBM (ICDAR 2019 Best Paper) | Layout detection: text, title, list, figure, table | 360K+ document images | mAP (COCO) | High — underpins Docling |
| **DocVQA** | WACV 2021 | Visual QA on document images | 50K questions, 12.7K images | ANLS | Medium |
| **PDF Extraction Benchmark** | Procycons 2025 | Practical comparison: Docling vs Unstructured vs LlamaParse | Practical test suite | Accuracy % | **Very High** — validates Docling choice |

### 3. Knowledge Graph Extraction Benchmarks

| Benchmark | Source | What It Measures | Size | Metrics | DocStill Relevance |
|-----------|--------|-----------------|------|---------|-------------------|
| **KGGen/MINE** | NeurIPS 2025 | KG extraction quality from text; % of facts captured | 100 articles, 15 facts each | Fact capture % (semantic sim + LLM judge) | **Very High** — exactly DocStill's task |
| **SciERC** | UW NLP (EMNLP 2018) | Multi-task scientific KG: NER + relations + coreference | 500 abstracts, 6 entity types, 7 relation types | F1 | **Very High** — concept types match DocStill |
| **SciREX** | Allen AI (ACL 2020) | Full-document IE from scientific papers, N-ary relations | 438 full papers | F1 | **Very High** — full-doc IE |
| **DocRED** | THUNLP (ACL 2019) | Document-level relation extraction, cross-sentence reasoning | 5,053 docs, 96 relation types | F1 (with Ign-F1) | **High** — multi-sentence reasoning |
| **Text2KGBench** | ISWC 2023 | Ontology-driven KG generation + hallucination detection | 18K+ sentences | 7 metrics incl. hallucination | **High** — measures faithfulness |
| **TACRED** | Stanford NLP | Sentence-level relation extraction | 106K examples, 41 relations | Micro F1 | Low-Medium |

### 4. Book-Specific & Educational KG Benchmarks

| Benchmark | Source | What It Measures | DocStill Relevance |
|-----------|--------|-----------------|-------------------|
| **ACE (EKG-Dataset)** | JEDM 2024 | Educational KG construction with prerequisite relations from textbooks | **Very High** — concept prerequisites |
| **Textbook to Triples** | arXiv 2111.10692 | KG triple extraction at sentence/paragraph/entity levels from textbook chapters | **Very High** — exact use case |
| **KnowEdu** | Heliyon 2024 review | Concept extraction (F1>0.70) + relation identification (AUC=0.95) from educational content | **High** |

### 5. Industry/Commercial Benchmarks

| Benchmark | Source | What It Measures | DocStill Relevance |
|-----------|--------|-----------------|-------------------|
| **FACTS Grounding** | Google DeepMind 2024-25 | LLM factual accuracy from documents (up to 32K tokens); hallucination detection | **High** — extraction faithfulness |
| **StructEval** | Tiger AI Lab 2025 | LLM structured output in 18 formats (JSON, YAML, CSV, etc.); GPT-4o only 76% | **High** — DocStill's output formats |
| **CRAG** | Meta 2024 | End-to-end RAG quality | Medium — downstream RAG usage |

### 6. Key Evaluation Metrics

| Metric | Used For | DocStill Application |
|--------|----------|---------------------|
| **Exact Match (EM)** | QA | Concept name accuracy |
| **Token-level F1** | QA, NER | Definition extraction quality |
| **Micro F1** | Relation extraction | Relationship accuracy |
| **ROUGE-1/2/L** | Summarization | Concept summary quality |
| **TEDS** | Table recognition | Table extraction accuracy |
| **mAP** | Layout detection | Layout element quality |
| **MINE Score** | KG extraction | Overall KG completeness |
| **Hallucination Rate** | Grounding | Faithfulness of extraction |

### 7. Gaps — No Existing Benchmark Covers

- Extracting **code patterns** and **implementation examples** from technical books
- Converting extracted knowledge into **agent-consumable formats** (CLAUDE.md, system prompts)
- Evaluating **hierarchy quality** (concept taxonomies, chapter-to-concept mappings)
- End-to-end from **raw PDF to structured KG** as a single pipeline
- Benchmarks using **real technical textbooks** (most use papers or Wikipedia)

**These gaps represent the opportunity for DocStill's custom benchmark.**

---

## Part 2: Custom Benchmark Design

### 1. Overview

A five-dimension benchmark that evaluates each pipeline stage independently:

| Dimension | What It Measures | Key Metrics |
|-----------|-----------------|-------------|
| **Parsing** | Text, tables, code blocks, figures, TOC extraction | TOC precision/recall, text fidelity, table accuracy |
| **Chunking** | Semantic coherence, boundary respect, token limits | Boundary respect rate, token compliance, heading preservation |
| **Concept Extraction** | Finding concepts, definitions, categories | Precision, recall (must-find), F1, definition quality |
| **Relationship Extraction** | Edges, types, directions | Precision, recall, F1, type accuracy, direction accuracy |
| **End-to-End** | Downstream usability for AI agents | Query answer coverage, list precision/recall, definition match |

### 2. Ground Truth Generation Strategy

**Bootstrapping with AI + SME validation:**

1. **Seed book**: Data Structures & Algorithms textbook
2. **AI generates candidates**: Claude produces TOC, concepts, definitions, relationships, query-answer pairs
3. **SME validates**: Domain expert reviews/corrects AI output, marks priority (must_find vs nice_to_have)
4. **Iterate**: Start with 1 chapter → expand to full book → add second domain

**Iteration plan:**
```
Iter 0: 1 chapter, ~10-15 concepts, ~5 queries
Iter 1: 2-3 chapters, ~30-40 concepts, ~15 queries
Iter 2: Full book (5 chapters), ~60-80 concepts, ~30 queries
Iter 3: Add second book (accounting), ~120+ concepts, ~50+ queries
```

### 3. Two-Tier Priority System

- **`must_find`**: Core concepts/relationships that extraction must catch. Missing = test failure.
- **`nice_to_have`**: Peripheral items that contribute to scores but don't cause failure.

This prevents the benchmark from being too lenient or too brittle.

### 4. Ground Truth JSON Schema

A comprehensive schema covering all five dimensions. Key types:

- **`ExpectedConcept`**: name, aliases, definition, category, priority, prerequisites, source_page, expected_confidence_min
- **`ExpectedRelationship`**: source, target, type (is_a, requires, part_of, uses, implements, ...), priority, justification
- **`QueryAnswerPair`**: query, query_type (list/definition/relationship/comparison), expected_answer, answer_is_exhaustive
- **`TOCEntry`**: title, level, page, children (hierarchical)

### 5. Example: DSA Chapter 2 — Sorting Algorithms

**Concepts (11 entries):**
- Must-find: Sorting Algorithm, Quick Sort, Merge Sort, Heap Sort, Bubble Sort, Pivot, Partitioning, Stability, Binary Heap
- Nice-to-have: In-Place Sorting, Heapify

**Relationships (12 entries):**
- Must-find: Quick Sort →is_a→ Sorting Algorithm, Merge Sort →is_a→ Sorting Algorithm, Quick Sort →implements→ Divide and Conquer, Heap Sort →uses→ Binary Heap, Quick Sort →uses→ Partitioning, etc.

**Queries (10 entries):**
- "List all sorting algorithms in Chapter 2" → [Bubble Sort, Quick Sort, Merge Sort, Heap Sort]
- "Which sorting algorithms have O(n log n) average case?" → [Quick Sort, Merge Sort, Heap Sort]
- "What data structure does Heap Sort depend on?" → Binary Heap
- "Compare Quick Sort and Merge Sort in terms of space complexity" → (detailed comparison)

### 6. Proposed File Layout

```
benchmarks/
  schema/
    benchmark_entry.schema.json
  ground_truth/
    dsa_ch2_sorting.json          # First ground truth file
    dsa_full.json                 # Full book (future)
    accounting_ch1.json           # Second domain (future)
  results/
    .gitkeep
  docstill_benchmark/
    __init__.py
    models.py                     # Pydantic models for ground truth + scores
    evaluate_parsing.py
    evaluate_chunking.py
    evaluate_concepts.py
    evaluate_relationships.py
    evaluate_e2e.py
    runner.py                     # CLI: load ground truth → run pipeline → evaluate
    report.py                     # Score reporting + comparison
  conftest.py
  test_benchmark_*.py             # pytest tests per dimension
```

### 7. Running the Benchmark

```bash
# Via pytest (per-dimension)
.venv/bin/python -m pytest benchmarks/ -q --override-ini="addopts=" 2>&1

# Via CLI runner (full pipeline)
.venv/bin/python -m benchmarks.docstill_benchmark.runner \
  --ground-truth benchmarks/ground_truth/dsa_ch2_sorting.json \
  --book-path data/books/raw/dsa_textbook.pdf \
  --output benchmarks/results/run_2026-02-09.json

# Compare runs
.venv/bin/python -m benchmarks.docstill_benchmark.report \
  --baseline benchmarks/results/run_2026-02-01.json \
  --current benchmarks/results/run_2026-02-09.json
```

### 8. Terminal Output Example

```
DocStill Benchmark Report: dsa_ch2_sorting v1.0.0
==================================================
Dimension           Score   Target  Status
------------------  ------  ------  ------
Parsing             0.96    0.95    PASS
Chunking            0.96    0.95    PASS
Concept Extraction  0.83    0.80    PASS
Relationships       0.81    0.70    PASS
End-to-End          0.81    0.80    PASS
--------------------------------------------------
Overall             0.87    0.80    PASS

Missed must-find concepts (1):
  - Heapify

Spurious concepts (2):
  - Array Traversal
  - Swap Operation
```

### 9. Implementation Priorities

| Priority | Task | Effort |
|----------|------|--------|
| P0 | JSON schema + first ground truth (DSA Ch2) | 1 day |
| P0 | `evaluate_concepts.py` (precision/recall/F1) | 1 day |
| P0 | `evaluate_relationships.py` | 1 day |
| P1 | `evaluate_parsing.py` (TOC + text snippets) | 0.5 day |
| P1 | `evaluate_chunking.py` | 0.5 day |
| P1 | `runner.py` CLI | 0.5 day |
| P2 | `evaluate_e2e.py` (query-answer) | 1 day |
| P2 | `report.py` (comparison) | 0.5 day |
| P2 | Second ground truth file (accounting) | 1 day |
| P3 | CI integration | 0.5 day |
| P3 | Time-series visualization | 1 day |

### 10. Key Design Decisions

1. **Human-validated ground truth** — AI generates candidates; SMEs correct them
2. **Two-tier priorities** — must_find (test failure) vs nice_to_have (score contribution only)
3. **Per-dimension isolation** — parsing bugs show in parsing score, not hidden by LLM compensation
4. **Reuse existing code** — `_token_overlap_similarity`, `score_extraction`, `RelationshipType`
5. **Mock-friendly** — runs with `DOCSTILL_LLM_MODE=mock` for CI (parsing/chunking only), full LLM locally
6. **Version-tracked** — ground truth has version field; results record which version they used
